Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/StubIr.kt

```diff
@@ -0,0 +1,460 @@
+
+interface StubIrElement {
+    fun accept(visitor: StubIrVisitor)
+
+    fun <T, R> accept(visitorAux: StubIrVisitorAux<T, R>, data: T): R
+}
+
+interface StubContainer : StubIrElement {
+    val meta: StubContainerMeta
+    val classes: List<ClassStub>
+    val functions: List<FunctionalStub>
+    val properties: List<PropertyStub>
+    val typealiases: List<TypealiasStub>
+    val simpleContainers: List<SimpleStubContainer>
+}
+
+/**
+ * Meta information about [StubContainer].
+ * For example, can be used for comments in textual representation.
+ */
+class StubContainerMeta(
+        val textAtStart: String = "",
+        val textAtEnd: String = ""
+)
+
+/**
+ * Trivial container of IR elements that can be used
+ * for module or ObjC category representation.
+ */
+class SimpleStubContainer(
+        override val meta: StubContainerMeta = StubContainerMeta(),
+        override val classes: List<ClassStub> = emptyList(),
+        override val functions: List<FunctionalStub> = emptyList(),
+        override val properties: List<PropertyStub> = emptyList(),
+        override val typealiases: List<TypealiasStub> = emptyList(),
+        override val simpleContainers: List<SimpleStubContainer> = emptyList()
+) : StubContainer {
+    override fun accept(visitor: StubIrVisitor) {
+        visitor.visitSimpleStubContainer(this)
+    }
+
+    override fun <T, R> accept(visitorAux: StubIrVisitorAux<T, R>, data: T): R {
+        return visitorAux.visitSimpleStubContainer(this, data)
+    }
+}
+
+val StubContainer.children: List<StubIrElement>
+    get() {
+        val classes: List<StubIrElement> = classes
+        val funcs: List<StubIrElement> = functions
+        return classes + properties + funcs + typealiases
+    }
+
+
+sealed class StubType {
+    abstract val nullable: Boolean
+}
+
+/**
+ * Marks that abstract value of such type can be passed as value.
+ */
+sealed class ValueStub
+
+class TypeParameterStub(
+        val name: String,
+        val upperBound: StubType? = null
+) {
+    fun getStubType(nullable: Boolean): StubType =
+            SymbolicStubType(name, nullable = nullable)
+
+}
+
+// Add variance if needed
+class TypeArgumentStub(val type: StubType)
+
+/**
+ * Wrapper over [KotlinType].
+ */
+class WrapperStubType(
+        val kotlinType: KotlinType
+) : StubType() {
+    override val nullable: Boolean
+        get() = when (kotlinType) {
+            is KotlinClassifierType -> kotlinType.nullable
+            is KotlinFunctionType -> kotlinType.nullable
+            else -> error("Unknown KotlinType: $kotlinType")
+        }
+}
+
+/**
+ * Wrapper over [Classifier].
+ */
+class ClassifierStubType(
+        val classifier: Classifier,
+        val typeArguments: List<TypeArgumentStub> = emptyList(),
+        override val nullable: Boolean = false
+) : StubType()
+
+
+/**
+ * Fallback variant for all cases where we cannot refer to specific [KotlinType] or [Classifier].
+ */
+class SymbolicStubType(
```

Comment:
Note that this type is not convertible to metadata.

Reply:
Yeah, `SymbolicStubType` is a hacky approach. It'll be more clear how to do it right when I'll start to work on metadata-based "backend".